app/src/test/java/org/simple/clinic/facility/change/FacilityChangeScreenControllerTest.kt

```diff
@@ -138,4 +175,102 @@ class FacilityChangeScreenControllerTest {
+  }
+
+  @Test
+  fun `when screen is started then location should only be read once`() {
+    configProvider.onNext(configTemplate.copy(locationListenerExpiry = Duration.ofSeconds(5)))
+
+    val facilities = listOf(
+        PatientMocker.facility(name = "Facility 1"),
+        PatientMocker.facility(name = "Facility 2"))
+    whenever(facilityRepository.facilitiesInCurrentGroup(any(), any())).thenReturn(Observable.just(facilities))
+
+    val timeSinceBootWhenRecorded = Duration.ofMillis(elapsedRealtimeClock.millis())
+    whenever(locationRepository.streamUserLocation(any(), any())).thenReturn(
+        Observable.just(
+            LocationUpdate.Available(Coordinates(0.0, 0.0), timeSinceBootWhenRecorded),
+            LocationUpdate.Unavailable,
+            LocationUpdate.Available(Coordinates(0.0, 0.0), timeSinceBootWhenRecorded)))
+
+    uiEvents.onNext(ScreenCreated())
+    uiEvents.onNext(FacilityChangeSearchQueryChanged(""))
+    uiEvents.onNext(FacilityChangeLocationPermissionChanged(GRANTED))
+
+    testComputationScheduler.advanceTimeBy(6, SECONDS)
+
+    verify(locationRepository).streamUserLocation(any(), any())
+    verify(screen, times(1)).updateFacilities(any(), any())
+  }
+
+  @Test
+  fun `when the user's location updates are received then only one recent update should be read`() {
+    val config = configTemplate.copy(staleLocationThreshold = Duration.ofMinutes(10))
+    configProvider.onNext(config)
+
+    val facilities = listOf(PatientMocker.facility(name = "Facility 1"))
+    whenever(facilityRepository.facilitiesInCurrentGroup(any(), any())).thenReturn(Observable.just(facilities))
+
+    val locationUpdates = PublishSubject.create<LocationUpdate>()
+    whenever(locationRepository.streamUserLocation(any(), any())).thenReturn(locationUpdates)
+
+    uiEvents.onNext(ScreenCreated())
+    uiEvents.onNext(FacilityChangeSearchQueryChanged(""))
+    uiEvents.onNext(FacilityChangeLocationPermissionChanged(GRANTED))
+
+    val locationOlderThanStaleThreshold = Available(
+        location = Coordinates(0.0, 0.0),
+        timeSinceBootWhenRecorded = Duration.ofMillis(elapsedRealtimeClock.millis()))
+
+    elapsedRealtimeClock.advanceBy(config.staleLocationThreshold + Duration.ofSeconds(1))
+
+    locationUpdates.onNext(locationOlderThanStaleThreshold)
+    verify(screen, never()).updateFacilities(any(), any())
+
+    locationUpdates.onNext(locationOlderThanStaleThreshold)
+    verify(screen, never()).updateFacilities(any(), any())
+
+    elapsedRealtimeClock.advanceBy(config.staleLocationThreshold)
+
+    val locationNewerThanStaleThreshold = Available(
+        location = Coordinates(0.0, 0.0),
+        timeSinceBootWhenRecorded = Duration.ofMillis(elapsedRealtimeClock.millis()))
+
+    locationUpdates.onNext(locationNewerThanStaleThreshold)
+    verify(screen).updateFacilities(any(), any())
+  }
+
+  @Suppress("unused")
+  fun `params for permission denials`(): List<RuntimePermissionResult> {
```

Comment:
Move this function to below the test it is being used in.

Reply:
Aye